src/modules/lyrics/injectLyrics.ts

```diff
@@ -242,40 +252,82 @@ export function injectLyrics(data: LyricSourceResultWithMeta, keepLoaderVisible
     const shouldRomanize =
       (data.language && Constants.romanizationLanguages.includes(data.language)) || containsNonLatin(item.words);
     const canInjectRomanizationsEarly = (shouldRomanize && item.romanization) || romanizedCacheResult !== null;
+    if (item.romanization) {
+      romanizedCacheResult = item.romanization;
+    }
 
     if (canInjectRomanizationsEarly) {
-      if (item.romanization) {
-        romanizedCacheResult = item.romanization;
-      }
-
       if (romanizedCacheResult !== item.words) {
-        createBreakElem(lyricElement, 4);
-
-        let romanizedLine = document.createElement("div");
-        romanizedLine.classList.add(Constants.ROMANIZED_LYRICS_CLASS);
-
         if (item.timedRomanization && item.timedRomanization.length > 0) {
-          createLyricsLine(item.timedRomanization, line, romanizedLine);
+          createLyricsLine(item.timedRomanization, line, createRomanizedElem());
         } else {
-          romanizedLine.textContent = "\n" + romanizedCacheResult;
+          createRomanizedElem().textContent = "\n" + romanizedCacheResult;
         }
-        romanizedLine.style.order = "5";
-        lyricElement.appendChild(romanizedLine);
-        lyricElement.dataset.romanized = "true";
       }
+    } else {
+      langPromise.then(source_language => {
+        Translation.onRomanizationEnabled(async () => {
+          let isNonLatin = containsNonLatin(item.words);
+          if (Constants.romanizationLanguages.includes(source_language) || isNonLatin) {
+            let usableLang = source_language;
+            if (isNonLatin && !Constants.romanizationLanguages.includes(source_language)) {
+              usableLang = "auto";
+            }
+
+            if (item.words.trim() !== "♪" && item.words.trim() !== "") {
+              let result;
+              if (item.romanization) {
+                result = item.romanization;
+              } else {
+                result = await Translation.translateTextIntoRomaji(usableLang, item.words);
+              }
+
+              if (result && !isSameText(result, item.words)) {
+                createRomanizedElem().textContent = result;
+                lyricsElementAdded();
+              }
+            }
+          }
+        });
+      });
     }
 
+    let createTranslationElem = () => {
+      createBreakElem(lyricElement, 6);
+      let translatedLine = document.createElement("div");
+      translatedLine.classList.add(Constants.TRANSLATED_LYRICS_CLASS);
+      translatedLine.style.order = "7";
+      lyricElement.appendChild(translatedLine);
+      return translatedLine;
+    };
+
     let translationResult: TranslationResult | null;
 
-    if (item.translation && Translation.getCurrentTranslationLanguage() === item.translation.lang) {
+    let currentTranslationLang = Translation.getCurrentTranslationLanguage();
+    // Compare base language codes, e.g. "en" matches "en-US"
+    function langCodesMatch(lang1: string, lang2: string): boolean {
```

Comment:
can we move this function outside `lyrics.lines.forEach((item: Lyric, lineIndex) => {`?
i believe right now this function gets recreated for every lyric line